tests/functional/test_install.py

```diff
@@ -1527,3 +1528,99 @@ def test_target_install_ignores_distutils_config_install_prefix(script):
+    assert result.returncode != 0
+    new_command = [sys.executable, '-m', 'pip'] + command[1:]
+    assert 'To modify pip, please run the following command:\n{}'.format(
+        ' '.join(new_command)) in result.stderr, str(result)
+
+
+@pytest.mark.skipif("sys.platform != 'win32'")
+def test_protect_pip_from_modification_via_deps_on_windows(script, with_wheel):
+    """
+    Test ``pip install pkga`` is raised an error on Windows
+    if `pkga` implicitly tries to upgrade pip.
+    """
+    # Make a wheel for pkga which requires pip
+    script.scratch_path.joinpath('pkga').mkdir()
+    pkga_path = script.scratch_path / 'pkga'
+    pkga_path.joinpath('setup.py').write_text(textwrap.dedent("""
+        from setuptools import setup
+        setup(name='pkga',
+              version='0.1',
+              install_requires = ["pip<{}"])
+    """.format(pip.__version__)))
+    script.run(
+        'python', 'setup.py', 'bdist_wheel', '--universal', cwd=pkga_path
+    )
+
+    # Make sure pip install pkga is raised error
+    pkga_wheel_path = './pkga/dist/pkga-0.1-py2.py3-none-any.whl'
+    command = ['pip', 'install', pkga_wheel_path]
+    result = script.run(*command, expect_error=True)
+    assert result.returncode != 0
+    new_command = [sys.executable, "-m"] + command
+    assert "To modify pip, please run the following command:\n{}".format(
+        " ".join(new_command)) in result.stderr, str(result)
+
+
+@pytest.mark.skipif("sys.platform != 'win32'")
+def test_protect_pip_from_modification_via_sub_deps_on_windows(
+    script, with_wheel
+):
+    """
+    Test ``pip install pkg`` is raised an error on Windows
+    if sub-dependencies of `pkg` implicitly tries to upgrade pip.
+    """
+    # Make a wheel for pkga which requires pip
+    script.scratch_path.joinpath('pkga').mkdir()
+    pkga_path = script.scratch_path / 'pkga'
+    pkga_path.joinpath('setup.py').write_text(textwrap.dedent("""
+        from setuptools import setup
+        setup(name='pkga',
+              version='0.1',
+              install_requires = ["pip<{}"])
+    """.format(pip.__version__)))
+    script.run(
+        'python', 'setup.py', 'bdist_wheel', '--universal', cwd=pkga_path
+    )
+
+    # Make a wheel for pkgb which requires pkga
+    script.scratch_path.joinpath('pkgb').mkdir()
+    pkgb_path = script.scratch_path / 'pkgb'
+    pkgb_path.joinpath('setup.py').write_text(textwrap.dedent("""
+        from setuptools import setup
+        setup(name='pkgb',
+              version='0.1',
+              install_requires = ["pkga"])
+    """))
+    script.run(
+        'python', 'setup.py', 'bdist_wheel', '--universal', cwd=pkgb_path
+    )
+
+    # Make sure pip install pkgb is raised an error
+    pkgb_wheel_path = './pkgb/dist/pkgb-0.1-py2.py3-none-any.whl'
+    command = [
+        'pip', 'install', pkgb_wheel_path, '--find-links', pkga_path / 'dist'
+    ]
+    result = script.run(*command, expect_error=True)
```

Comment:
It's preferable to use `script.pip_install_local` if we're not using the network.

Further, having the command inline avoids the need for additional variables etc, which simplifies the call invocation.

```suggestion
    result = script.pip_install_local(
        pkgb_wheel_path, '--find-links', pkga_path / 'dist',
        expect_error=True,
    )
```

Reply:
That's a good point, but not in this scenario, because pkga depends on pip and needs to upgrade it. That's why we need network here, see:

```
E       AssertionError: Script result: python -m pip install --no-index --find-links file:///Users/albert/Projects/pip/tests/data/packages /private/var/folders/4x/rz_w89dx1z3bydfd6qpdjn_h0000gn/T/pytest-of-albert/pytest-35/test_protect_pip_from_modifica0/workspace/scratch/pkga-0.1-py2.py3-none-any.whl
E           return code: 1
E         -- stderr: --------------------
E         ERROR: Could not find a version that satisfies the requirement pip!=19.3.dev0 (from pkga==0.1) (from versions: test-package-0.1, test-package-0.1.1)
E         ERROR: No matching distribution found for pip!=19.3.dev0 (from pkga==0.1)
```

Reply:
Right, then let's use `script.pip`. :)

Reply:
So i'll resolve this and the other.

Reply:
Wait, you mean change `script.run` to `script.pip`?

Reply:
Okay, fixed to `script.pip`.